pkg/printers/customcolumn_flags.go

```diff
@@ -0,0 +1,121 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package printers
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/spf13/cobra"
+
+	"k8s.io/kubernetes/pkg/kubectl/scheme"
+)
+
+// CustomColumnsPrintFlags provides default flags necessary for printing
+// custom resource columns from an inline-template or file.
+type CustomColumnsPrintFlags struct {
+	NoHeaders        *bool
+	TemplateArgument *string
+}
+
+// ToPrinter receives an outputFormat and returns a printer capable of
+// handling custom-column printing.
+// Returns false if the specified outputFormat does not match a supported format.
+// Supported format types can be found in pkg/printers/printers.go
+func (f *CustomColumnsPrintFlags) ToPrinter(outputFormat string) (ResourcePrinter, bool, error) {
+	if len(outputFormat) == 0 {
+		return nil, false, fmt.Errorf("missing output format")
+	}
+
+	templateArg := ""
+
+	supportedFormats := map[string]bool{
+		"custom-columns-file": true,
+		"custom-columns":      true,
+	}
+
+	if f.TemplateArgument == nil || len(*f.TemplateArgument) == 0 {
+		for format := range supportedFormats {
+			format = format + "="
+			if strings.HasPrefix(outputFormat, format) {
+				templateArg = outputFormat[len(format):]
+				outputFormat = format[:len(format)-1]
+				break
+			}
+		}
+	} else {
+		templateArg = *f.TemplateArgument
+	}
+
+	if _, supportedFormat := supportedFormats[outputFormat]; !supportedFormat {
+		return nil, false, nil
+	}
+
+	if len(templateArg) == 0 {
+		return nil, true, fmt.Errorf("custom-columns format specified but no custom columns given")
+	}
+
+	decoder := scheme.Codecs.UniversalDecoder()
+
+	if outputFormat == "custom-columns-file" {
+		file, err := os.Open(templateArg)
+		if err != nil {
+			return nil, true, fmt.Errorf("error reading template %s, %v\n", templateArg, err)
+		}
+		defer file.Close()
+		p, err := NewCustomColumnsPrinterFromTemplate(file, decoder)
+		if err != nil {
+			return nil, true, err
+		}
+
+		return p, true, nil
+	}
+
+	noHeaders := false
+	if f.NoHeaders != nil {
+		noHeaders = *f.NoHeaders
+	}
+
+	p, err := NewCustomColumnsPrinterFromSpec(templateArg, decoder, noHeaders)
+	return p, true, err
+}
+
+// AddFlags receives a *cobra.Command reference and binds
+// flags related to custom-columns printing
+func (f *CustomColumnsPrintFlags) AddFlags(c *cobra.Command) {
+	if f.NoHeaders != nil {
+		c.Flags().BoolVar(f.NoHeaders, "no-headers", *f.NoHeaders, "When using the default or custom-column output format, don't print headers.")
+	}
+
+	if f.TemplateArgument != nil {
+		c.Flags().StringVar(f.TemplateArgument, "template", *f.TemplateArgument, "Template string or path to template file to use when -o=go-template, -o=go-template-file. The template format is golang templates [http://golang.org/pkg/text/template/#pkg-overview].")
```

Comment:
@soltysh @deads2k the `--template` flag is also bound by `KubeTemplatePrintFlags` here: https://github.com/kubernetes/kubernetes/pull/61304/files#diff-6ee7dad6fdeccf4e8fe8669b7080b962R95

A solution could be to entirely drop support for the `--template` flag when printing custom-columns, and expect users to specify a custom-columns template or template file via the `--output` flag: `--output=custom-columns=NAME:.metadata.name`

Reply:
We can't do that, we need to maintain backwards compatibility.

Reply:
Usually when talking about template we allow all three of them. Maybe we need an upper level struct that contains all 3 (custom, go-template and jsonpath) so they can share a single `--template` flag.